fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -414,6 +414,33 @@ def module(self) -> FlattenParamsWrapper:
         assert isinstance(self._fsdp_wrapped_module, FlattenParamsWrapper)
         return self._fsdp_wrapped_module
 
+    def append_shared_param(self, p: Parameter) -> None:
+        """ Add a param that's already owned by another FSDP wrapper.
+
+            .. warning:: This is experimental!
+
+            This only works with all sharing FSDP modules are un-flattened.
+
+            p must to be already sharded by the owning module.
+
+            Check the corresponding unit test to see how is it used and tested.
+            In particular, the sharing FSDP wrappers are "siblings" not "parent"
+            and "child" of each other in the nested module structure.
+
+        Args:
+            p (Parameter):
+                The shared parameter.
+        """
+        assert self._is_root is None
+        assert not self.flatten_parameters
+        assert isinstance(p, Parameter)
+        assert p._is_sharded
+        p._is_shared = True
+        assert (
+            len(list(filter(lambda p: not (hasattr(p, "_is_shared") and p._is_shared), self.params))) > 0
+        ), "Must have at least 1 non-shared param."
```

Comment:
what does this requirement satisfy?

Reply:
This is due to autograd’s hook firing order